pkg/controller/utils/utils.go

```diff
@@ -1082,22 +1082,64 @@ func MaintainInstallationFinalizer(
 
 		// Check if the namespaced secondaryResources are still present.
 		// Keep track of all the secondary resources that the main resource creates.
-		// Only delete the finalizer if all of the secondary resources are deleted.
+		// Only delete the finalizer if all of the secondary resources are deleted and there are no lingering Pods.
 		for _, secondaryResource := range secondaryResources {
 			err := c.Get(ctx, types.NamespacedName{Namespace: secondaryResource.GetNamespace(), Name: secondaryResource.GetName()}, secondaryResource)
 			if err != nil && !errors.IsNotFound(err) {
 				return err
 			} else if errors.IsNotFound(err) {
-				log.Info("Object no longer exists.", "object", secondaryResource)
-			} else {
-				log.Info("Object is still present, waiting for termination", "object", secondaryResource)
+				log.V(2).Info("Object no longer exists.", "object", secondaryResource)
+			} else if err == nil {
+				log.V(2).Info("Object is still present, waiting for termination", "object", secondaryResource)
 				return nil
 			}
+
+			// If the secondary resource itself is gone, ensure there are no Pods left over that would still rely on CNI.
+			// Only applicable for workload types that create Pods.
+			switch obj := secondaryResource.(type) {
+			case *appsv1.Deployment:
+				if exists, count, err := hasLingeringPods(ctx, c, obj.Namespace, obj.Spec.Selector.MatchLabels); err != nil {
+					return err
+				} else if exists {
+					log.V(2).Info("Lingering Pods still exist for Deployment", "deployment", obj.Name, "count", count)
+					return nil
+				}
+			case *appsv1.DaemonSet:
+				if exists, count, err := hasLingeringPods(ctx, c, obj.Namespace, obj.Spec.Selector.MatchLabels); err != nil {
+					return err
+				} else if exists {
+					log.V(2).Info("Lingering Pods still exist for DaemonSet", "daemonset", obj.Name, "count", count)
+					return nil
+				}
+			case *appsv1.StatefulSet:
+				if exists, count, err := hasLingeringPods(ctx, c, obj.Namespace, obj.Spec.Selector.MatchLabels); err != nil {
+					return err
+				} else if exists {
+					log.V(2).Info("Lingering Pods still exist for StatefulSet", "statefulset", obj.Name, "count", count)
+					return nil
+				}
+			default:
+				// Non-workload resource; no Pod listing needed.
+			}
 		}
-		log.Info("All objects no longer exist. Removing finalizer", "finalizer", finalizer)
+		log.V(2).Info("All objects and their Pods no longer exist. Removing finalizer", "finalizer", finalizer)
```

Comment:
ah maybe i shouldn't've .V(2)'d this one

Reply:
Yeah, I'd prefer we leave all of these logs at regular verbosity - they are rare and only occur during teardown, which is unfortunately an error prone process and having the logs at all verbosity levels has proven helpful!